lua/bson_lua_wrapper.c

```diff
@@ -0,0 +1,496 @@
+    if (lua_istable(L, -1) == 0) {
+      snprintf(errorMessage, 255, "Table is not in BSON format (key: {type, value})");
+      bson_array_deinitialize(arr);
+      return -1;
+    }
+
+    const int result = bson_put_map_value(L, arr, NULL, errorMessage); //Stack: [bson_array]
+    if (result != 0) {
+      bson_array_deinitialize(arr);
+      return -1;
+    }
+    lua_pop(L, 1);
+  }
+  return 0;
+}
+
+/*
+  @brief Convert a Lua table into a BSON object.
+    Lua arguments:
+      1 (table) - Explicitly typed Lua table entry to be converted to BSON format.
+        This table should be in the following format:
+          {
+            type = element_type,
+            value = val
+          }
+
+  @param L - Lua state, used to modify the Lua stack
+  @param bson - The BSON object/array where the converted table is stored
+  @param errorMessage - To be populated with an error message if an error occurs
+
+  @return 0 on success, non-zero on failure
+*/
+static int bson_put_map_value(lua_State *L, void *bson, const char *key, char *errorMessage) {
+  //Retrieve the type of the entry and place it on the stack
+  lua_getfield(L, -1, "type"); //Stack: [{type, value}, type]
+
+  //Store the type of the entry and remove it from the stack
+  const lua_Integer type = luaL_checkinteger(L, -1);
+  lua_pop(L, 1); //Stack: [{type, value}]
+
+  //Retrieve the value of the entry and place it on the stack
+  lua_getfield(L, -1, "value"); //Stack: [{type, value}, value]
+  switch ((const element_type)type) {
+    case TYPE_DOCUMENT: {
+      BsonObject value;
+      const int result = table_to_bson_object(L, &value, errorMessage);
+      if (result != 0) {
+        return result;
+      }
+
+      //Assume array if no key is provided
+      if (key == NULL) {
+        bson_array_add_object((BsonArray *)bson, &value);
+      }
+      else {
+        bson_object_put_object((BsonObject *)bson, key, &value);
+      }
+      break;
+    }
+    case TYPE_ARRAY: {
+      BsonArray value;
+      const int result = table_to_bson_array(L, &value, errorMessage);
+      if (result != 0) {
+        return result;
+      }
+
+      //Assume array if no key is provided
+      if (key == NULL) {
+        bson_array_add_array((BsonArray *)bson, &value);
+      }
+      else {
+        bson_object_put_array((BsonObject *)bson, key, &value);
+      }
+      break;
+    }
+    case TYPE_INT32: {
+      lua_Integer value = luaL_checkinteger(L, -1);
+
+      //Assume array if no key is provided
+      if (key == NULL) {
+        bson_array_add_int32((BsonArray *)bson, (int32_t)value);
+      }
+      else {
+        bson_object_put_int32((BsonObject *)bson, key, (int32_t)value);
+      }
+      break;
+    }
+    case TYPE_INT64: {
+      lua_Integer value = luaL_checkinteger(L, -1);
+
+      //Assume array if no key is provided
+      if (key == NULL) {
+        bson_array_add_int32((BsonArray *)bson, (int64_t)value);
```

Comment:
@jacobkeeler Please use `bson_array_add_int64()` function instead `bson_array_add_int32()` here,
and `bson_object_put_int64()` instead `bson_object_put_int32()` in `else` statement in line 247 to prevent logical mistake.